Add Bidirectional Isolation section to formatting (#315)
unicode-org/message-format-wg: spec/formatting.md

It is an error for a local variable definition to
refer to a local variable that's defined after it in the message.

## Bidirectional Isolation

> Where appropriate,

Reviewer: I agree that "Where appropriate ... " is far too vague to be a conformance MUST. 

However, condition is too strong. 
1. There are many instances where isolation is not required, such as a single Arabic word in a placeholder surrounded by Latin. It would be better to supply a functional description. Off the top of my head: w
    - Where the text of the placeholder or surrounding text would rearrange due to the insertion.
2. There is a significant number of products in use that do not yet support isolates. [Aside: I really wish I'd made the embeddings be isolating in the first place, but water far under the bridge.] For use of MF2.0 with those products, we have to have a fallback position (ugly as that is).

> Where appropriate,

Author: I've updated the text following @aphillips's suggestions.

@macchiati Is there an existing algorithm for determining when isolation would not be required? I presume that this would require scanning both the placeholder's stringified value as well as that of its surroundings, and ensuring that neither have any weakly typed characters immediately adjacent to the other.

I would prefer an algorithm that would not require such introspection, even if it does come at a cost of needing to include isolating characters in some situations where they have no effect.

To support products and situations which do not support isolates, would it be sufficient to mandate the default behaviour, and allow for a formatter to provide alternatives? For example, in the JS implementation, I could imagine including an option like:

```ts
type MesageFormatOptions = {
  isolate = 'isolate' | 'embed' | 'none' = 'isolate',
  ...
}
```

This would allow for defaulting to isolation (explicitly spec-defined), while also being able to provide implementation-defined fallback solutions.

> the formatted string representation of a message MUST isolate message parts

Reviewer 2: "the formatted string representation" doesn't say what happens with formatToParts. Do message segments have lang/dir metadata or do they include isolating controls? I'd prefer metadata (so that one could implement markup instead of controls)

Reviewer 3: +1 for documenting the parts behavior.
I also prefer metadata, or special parts.
Whatever the "format to parts" will produce, but not raw Unicode bidi controls.

Reviewer: For the 'formatToString' functionality, I think we have no choice but to deploy the bidi controls. For the full data model, I agree that metadata is better. (One can argue that the formatToString is just semantic sugar, but I think the _end result_ of processing for formatToString() had better agree with that.)

Author: I've left out the formatted parts representation from this PR because I think that's a separate concern, and a topic on which I find us to have less agreement. In fact, I would be (pleasantly!) surprised if we were able to define a common formatted-part definition that could be shared by all implementations, given how much the design of such an implementation is tied to its surroundings.

Rather than discussing this further here, we should continue in #41 or #272.

> When a message part has a direction different from the message as a whole
> or when the part's direction is not known,

Author: ```suggestion
When a message part has a direction different from the message as a whole,
```

@macchiati With reference to your comment https://github.com/unicode-org/message-format-wg/pull/315#discussion_r1021884363, would this be the sort of change you're looking for? Or have I perhaps understood your "minimum" incorrectly?

The intent here would be to reduce the complexity required of implementations, such that they only need to add isolates if they explicitly know that a message and a part have different directionality. For instance, if an instance doesn't have the capability to determine the directionality of a variable `$foo`, it would not be required to isolate it.

To be clear, my own preference would be to require isolation by default in that case, but to allow for an implementation to be configurable to act differently, isolating either less, more, or somehow differently than defined in the spec.

> Where appropriate,
> the formatted string representation of a message MUST isolate message parts
> which may have different directionality than the message as a whole.
> Such a part MUST be prefixed with an explicit isolate character:

Author: ```suggestion
When a message part has a direction different from the message as a whole
or when the part's direction is not known,
the formatted string representation of the message MUST isolate the part in question
with an explicit isolate character:
```

- U+2066 LEFT-TO-RIGHT ISOLATE if the part is known to have LTR directionality,
- U+2067 RIGHT-TO-LEFT ISOLATE if the part is known to have RTL directionality, or
- U+2068 FIRST STRONG ISOLATE if the part's directionality is not certain.

> In all cases,

Author: Good point. That was my intent here.

```suggestion
In all cases where a part is prefixed with such an isolate character,
```

It is an error for a local variable definition to
refer to a local variable that's defined after it in the message.

## Bidirectional Isolation

Formatted messages may be bidirectional,
i.e. contain a mix of left-to-right and right-to-left content.
In all output formats,
such parts should be isolated from each other so that
the directionality of one part of the message does not affect

> the presentation of the rest of the message.

Reviewer 2: I think this doesn't quite explain the issue. Perhaps:

> Formatted messages can be bidirectional, that is, containing a mixture of left-to-right and right-to-left text. When replacing placeholders with formatted string values, the [Unicode Bidirectional Algorithm](http://www.unicode.org/reports/tr9/) \[UAX9\] can produce unexpected or undesirable effects, such as "spillover". In such cases, placeholders _should_ (must?) be bidirectionally "isolated" so that the directionality of the text in the placeholder does not negatively affect the presentation of the overall string.

We might also link to some of the work Richard and I have been doing, e.g. https://www.w3.org/International/articles/lang-bidi-use-cases/#bidiCase7

Author: ```suggestion
Formatted messages can be bidirectional,
containing a mixture of left-to-right and right-to-left text.
When replacing placeholders with formatted string values,
the [Unicode Bidirectional Algorithm](http://www.unicode.org/reports/tr9/) [UAX9]
can produce unexpected or undesirable effects, such as "spillover".
In such cases, placeholders should be bidirectionally "isolated"
so that the directionality of the text in the placeholder
does not negatively affect the presentation of the overall string.
```

I think it does need to be "should", because "must" would preclude an implementation from providing any isolation strategy that carries any risk of not isolating parts. And that feels like a bit of an overreach.

Btw, I'm using the leftmost "Add a suggestion" button of the comment editing toolbar here to enter a committable suggestion.

Implementations may provide one or more strategies for isolation,
but one such strategy must behave as follows,

> when formatting the message as a string:

Reviewer 2: I would reverse the normative statements for clarity to implementers:

> Implementations _must_ provide at least one isolation mechanism from the list below. They _may_ provide additional mechanisms in order to allow better overall control from users.

Author: At the moment, the PR only presents on mechanism/strategy. If we add the compatibility mode I mention in an earlier comment, I would prefer either requiring both or at least the compatibility mode of all implementations, rather than "at least one".

Reviewer 4: I also think that there's something not right in how this is current phrased. `MAY provide one or more...` and `one MUST behave as follows...` seem to be at odds with each other? Can an implementation provide zero strategies?

I liked the wording suggested by @macchiati in https://github.com/unicode-org/message-format-wg/pull/315#discussion_r1044690668. @aphillips's suggestion above also sounds good to me.

Considering a "formatted part" to be the formatted string representation of a Placeholder,
when the formatted part has a direction different from the message as a whole
or when its directionality is not known,
the formatted string representation of the message must

> prefix the formatted part with one of the following isolate characters:

Reviewer 2: This leaves a gap for same-base-paragraph-direction placeholders with various edge effects (generally starting or ending with weak runs or runs with neighboring character issues). I admit that we've wrapped ourselves up trying to define when we are required to isolate without having to run the bidi algorithm. I don't have a solution yet, so perhaps keeps this for now, but add a note about the issue??

I would split the term definition of "formatted part" from the requirement.

Author: I don't think I see the hole you're describing. Could you give an example?

Reviewer 2: The pattern string is:

> السعر {$price} + {$shipping_amt} الشحن

The currency code for both is `USD` and the symbol for that is `USD` in the `ar-AE` locale. The resulting string is:

> 
السعر 1,234.56 AED + 12.99 USD الشحن

It should be:

> السعر ⁧1,234.56 USD⁩ + ⁧12.99 USD⁩ الشحن

Here are the code point sequences with and without RLI/PDI around the prices:

```
without: \u0627\u0644\u0633\u0639\u0631 1,234.56 USD + 12.99 USD \u0627\u0644\u0634\u062d\u0646
with: \u0627\u0644\u0633\u0639\u0631 \u20671,234.56 USD\u2069 + \u206712.99 USD\u2069 \u0627\u0644\u0634\u062d\u0646
```

In both cases the base paragraph direction of the pattern string *and also* each of the placeables is RTL, but the appearance of weak (numbers) and strong (letters) LTR runs confuses bidi unless isolated.

Dates often have this sort of thing too or any string that starts or ends with numeric sequences or enclosing punctuation.

Author: Ok, I see it now, how the formatted value might be nominally RTL but only composed of neutral and LTR characters. It's interesting to play around with this in different browsers, as I see this:

```js
const usd = new Intl.NumberFormat('ar-AE', { style: 'currency', currency: 'USD'}).format(12.99)

// in Chrome & Node.js
usd === "US$ 12.99"
usd.split('').map(c=>c.charCodeAt(0))
// [ 85, 83, 36, 160, 49, 50, 46, 57,  57 ]

// in Firefox
usd === "‏12.99 US$"
usd.split('').map(c=>c.charCodeAt(0))
// [ 8207, 49, 50, 46, 57, 57, 160, 85, 83, 36 ]
```

Specifically, Firefox prefixes the string with an RTL mark for currencies. It doesn't really help with the isolation, though.

So I agree that this situation is left outside the "must" statement, I think it's potentially ok? As in, at least we don't have a "must not" so that an implementation is allowed to introspect the formatted part and determine that it ought to be isolated.

- U+2066 LEFT-TO-RIGHT ISOLATE if the formatted part is known to have LTR directionality,
- U+2067 RIGHT-TO-LEFT ISOLATE if the formatted part is known to have RTL directionality, or
- U+2068 FIRST STRONG ISOLATE if the formatted part's directionality is not known.

In all cases where a formatted part is prefixed with such an isolate character,
it must be postfixed with a corresponding U+2069 POP DIRECTIONAL ISOLATE character.

An implementation must define its own method for
determining the directionality of a formatted part,
relying either on metadata attached to the formatted part,

> or on introspection of its contents.

Reviewer 2: I would add one more callout here. Many of the formatted parts are produced by a formatter with a given locale. APIs exist in at least a few platforms for determining the default base direction based on locale. So perhaps:

> An implementation _must_ determine the base paragraph direction for each formatted part. These might include a metadata field/method, introspection of the formatted part using UBA \[UAX9\], or from the locale of the placeholder's formatter (for example, `Intl.Locale`'s [textInfo](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/textInfo) in JavaScript). Formatters are responsible for using whatever mechanism is provided by the implementation to provide appropriate directional information for each formatted part.

For example, an implementation could provide a `:number` formatting function
which would always produce output matching the message's locale,

> allowing for its formatted string representation to never need isolation.

Reviewer 2: You're assuming that the message formatter's locale matches the language of the pattern, which is a bad assumption. Many (most!) resource systems fall back or allow sparse population of translated resources. As a result, the locale of the message formatter (and its subformatters) may not match that of the pattern string. _This is a feature._ It is often the case that developers will wish to test code in a given locale before the translations are available or to use pseudo.

I also note that this paragraph is not an example of the preceding requirement. Perhaps insert this just ahead of this paragraph:

> Implementations _should not_ isolate formatted parts when the base paragraph direction of the pattern and the formatted parts match *and* there are no opposite direction runs of text in the formatted part.

Author: Let me give an example to verify if I've understood your concern right:

An app using English as its fallback locale is partially localized to Arabic, with some English messages filling in holes. If such a LTR message is formatted using an RTL formatter _and_ it contains a placeholder with a variable reference to an RTL value, we should isolate the placeholder's value in the output, even though the formatter and the value are both RTL.

If this is the case, then I think what we need is a more explicit definition somewhere of the "message locale", which I would consider to be English in the above example, even if the formatter is nominally in Arabic.

Reviewer 2: That's more or less effectively what I'm saying. The challenge is that (let's use Java as an example), I have code that looks like this:

```
ResourceBundle myPatterns = ResourceBundle.getBundle("com.example.Patterns", myRuntimeLocale);
MessageFormat2 formatter = new MessageFormat2(myPatterns.getString("somePattern"), myRuntimeLocale);
```

If you query `myPatterns` for its locale, you get `myRuntimeLocale`, even though the bundle is really a stack of bundles. If `myRuntimeLocale` is `ar-AE`, the actual translations probably live in the `ar` bundle. If the developer has added a new string, though, `somePattern` may be coming from the `ROOT` locale bundle (and be in, e.g. English). Both `formatter` and its subformat patterns are trying to serve the `ar-AE` locale, though and finding out which specific locale bundle coughed up the pattern is both inaccessible to it (MF2 doesn't know that the pattern is coming from a bundle--it's just a string, as far as it is concerned) and doesn't add any value to instrument in most cases.

The actual language (as opposed to the asserted language/locale of it) is not really a concern for MF2, as long as the pattern is valid. If we wanted to do language-sensitive processing (capitalization, for example), we'd still use a locale for that. 

When it comes to bidi, it also doesn't matter as much as the base paragraph direction does, save that the language might serve as a hint for what the direction should be (lacking other information).

It is an error for a local variable definition to
refer to a local variable that's defined after it in the message.

## Bidirectional Isolation

Formatted messages can be bidirectional,
containing a mixture of left-to-right and right-to-left text.
When replacing placeholders with formatted string values,
the [Unicode Bidirectional Algorithm](http://www.unicode.org/reports/tr9/) [UAX9]
can produce unexpected or undesirable effects, such as "spillover".
In such cases, placeholders should be bidirectionally "isolated"
so that the directionality of the text in the placeholder
does not negatively affect the presentation of the overall string.

Implementations may provide one or more strategies for isolation,
but one such strategy must behave as follows,
when formatting the message as a string:

> Considering a "formatted part" to be the formatted string representation of a Placeholder,

Reviewer 4: I suggest we not use `part` to mean the stringified result of formatting. We've sort of established `formatToParts()` as the name of the method which returns structured formatting results, which implies that a _part_ is one such structured result of formatting. Perhaps _formatted elements_ or _formatted pattern elements_ would work better?

4. The pattern `{Exact match}` of the most preferred `1` variant is selected.

## Handling Bidirectional Text

> Formatted messages MAY be bidirectional,
> containing a mixture of left-to-right and right-to-left spans of text.

Reviewer 2: This is inappropriately normative.

```suggestion
_Messages_ contain text which can be bidirectional, that is
consisting of a mixture of left-to-right and right-to-left spans of text.
```
